java/code/src/com/redhat/rhn/frontend/xmlrpc/saltkey/SaltKeyHandler.java

```diff
@@ -37,25 +40,120 @@ public SaltKeyHandler(SaltKeyUtils saltKeyUtilsIn) {
+
+    }
+
+
+    /**
+     * API endpoint to list pending salt keys
+     * @param loggedInUser the user
+     * @return 1 on success
+     *
+     * @xmlrpc.doc List pending salt keys
+     * @xmlrpc.param #param("string", "loggedInUser")
+     * @xmlrpc.returntype #array_single("string", "Pending salt key list")
+     */
+    public List<String> pendingList(User loggedInUser) {
+        try {
+            ensureOrgAdmin(loggedInUser);
+        }
+        catch (PermissionException e) {
+            throw new PermissionCheckFailureException(e);
+        }
+        return saltKeyUtils.unacceptedSaltKeyList(loggedInUser);
+    }
+
+    /**
+     * API endpoint to accept minion keys
+     * @param loggedInUser the user
+     * @param minionId the key identifier (minionId)
+     * @return 1 on success
+     *
+     * @xmlrpc.doc Accept a minion key
+     * @xmlrpc.param #param("string", "loggedInUser")
+     * @xmlrpc.param #param("string", "minionId")
+     * @xmlrpc.returntype #return_int_success()
+     */
+    public int accept(User loggedInUser, String minionId) {
+        try {
+            ensureOrgAdmin(loggedInUser);
+            saltKeyUtils.acceptSaltKey(loggedInUser, minionId);
+        }
+        catch (PermissionException e) {
+            throw new PermissionCheckFailureException(e);
+        }
+        catch (IllegalArgumentException e) {
+            throw new UnsupportedOperationException(e.getMessage());
+        }
+        return 1;
+    }
+
+
+    /**
+     * API endpoint to generate and accept minion keys
+     * @param loggedInUser the user
+     * @param minionId the key identifier (minionId)
+     * @return 1 on success
+     *
+     * @xmlrpc.doc Generate and accept a minion key
+     * @xmlrpc.param #param("string", "loggedInUser")
+     * @xmlrpc.param #param("string", "minionId")
+     * @xmlrpc.returntype #return_int_success()
+     */
+    public int genAccept(User loggedInUser, String minionId) {
+        try {
+            ensureOrgAdmin(loggedInUser);
+            saltKeyUtils.genAcceptSaltKey(loggedInUser, minionId);
+        }
+        catch (PermissionException e) {
+            throw new PermissionCheckFailureException(e);
+        }
+        catch (IllegalArgumentException e) {
+            throw new UnsupportedOperationException(e.getMessage());
+        }
+        return 1;
+    }
+
     /**
      * API endpoint to delete minion keys
      * @param loggedInUser the user
      * @param minionId the key identifier (minionId)
      * @return 1 on success
      *
      * @xmlrpc.doc Delete a minion key
-     * @xmlrpc.param #param("string", "sessionKey")
+     * @xmlrpc.param #param("string", "loggedInUser")
      * @xmlrpc.param #param("string", "minionId")
      * @xmlrpc.returntype #return_int_success()
      */
     public int delete(User loggedInUser, String minionId) {
-        ensureOrgAdmin(loggedInUser);
         try {
+            ensureOrgAdmin(loggedInUser);
```

Comment:
Same as above. This can live outside of the try/catch. Not sure if acceptSaltKey() can throw a PermissionException which needs to be converted. It might be possible when salt master cannot write some files.